compaction/compaction.cc

```diff
@@ -1386,6 +1386,20 @@ class split_compaction final : public regular_compaction {
     std::string_view report_finish_desc() const override {
         return "Split";
     }
+
+    virtual compaction_writer create_compaction_writer(const dht::decorated_key& dk) override {
+        auto sst = _sstable_creator(this_shard_id());
+        setup_new_sstable(sst);
+
+        auto token_range_of_new_sst = _table_s.get_post_split_token_range_of(dk.token());
+        const auto estimated_keys = _sstables[0]->estimated_keys_for_range(token_range_of_new_sst);
+
```

Comment:
Nice. did you test it manually? A simple test is to populate a tablet table and wait for splits. --target-tablet-size-in-bytes 20971520 shorten the time for splits to happen.

Reply:
I tested them in parts as I was not able to successfully trigger a split. I'll use your method and test it before submitting the next version.

Reply:
By reducing the `target-tablet-size-in-bytes`, I was able to trigger multiple splits on an sstable with 50K partitions. 

Without this fix, the split was invoked seven times before tablets were finally balanced and bloom filter was rebuilt twice for every split (14 rebuilds).
With the fix, the rebuild came down to 7 times in my attempts for the same number of splits.

Log of a test from when the sstable was flushed until the tablets were balanced : 

Note : There are multiple WARN messages in the log that emit the `estimated keys` value calculated by the patch - they are there because I was trying to get insight into the calculation and they are not a part of the final patch. 

```
INFO  2024-10-22 20:57:38,075 [shard 0:strm] api - perform_keyspace_flush: keyspace=ks tables=["test"]
INFO  2024-10-22 20:58:26,436 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311227
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Emitting resize decision of type split for table 2984a460-908a-11ef-93f4-09ea64b31b3e due to avg tablet size of 2311227
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:58:26,438 [shard 0: gms] raft_topology - updating topology state: Starting tablet migration
INFO  2024-10-22 20:58:26,457 [shard 0: gms] raft_topology - entered `tablet migration` transition state
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311227
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 20:58:26,457 [shard 0: gms] raft_topology - updating topology state: Finished tablet migration
INFO  2024-10-22 20:58:26,457 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16y2_0gi4w28zy0sfxww666-big-Data.db:level=0:origin=memtable]
WARN  2024-10-22 20:58:26,458 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Data.db` : 25024
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311227
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 20:58:26,591 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Filter.db: resizing bitset from 31288 bytes to 31336 bytes. sstable origin: split
WARN  2024-10-22 20:58:26,614 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_3nlnk28zy0sfxww666-big-Data.db` : 25831
INFO  2024-10-22 20:58:26,761 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_3nlnk28zy0sfxww666-big-Data.db:level=0]. 2MB to 2MB (~100% of original) in 298ms = 7MB/s. ~50048 total partitions merged to 49999.
INFO  2024-10-22 20:58:26,762 [shard 0:strm] table - Setting split ready sequence number to 1 for table ks.test
INFO  2024-10-22 20:59:26,437 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311657
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 1
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:59:26,438 [shard 0: gms] raft_topology - updating topology state: Started tablet split finalization
INFO  2024-10-22 20:59:26,456 [shard 0: gms] raft_topology - entered `tablet split finalization` transition state
INFO  2024-10-22 20:59:26,456 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
INFO  2024-10-22 20:59:26,456 [shard 0: gms] raft_topology - updating topology state: advance fence version to 6
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311657
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 1
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Split tablets for table 2984a460-908a-11ef-93f4-09ea64b31b3e, increasing tablet count from 1 to 2
INFO  2024-10-22 20:59:26,463 [shard 0: gms] raft_topology - updating topology state: Finished tablet split finalization
INFO  2024-10-22 20:59:26,469 [shard 0: gms] table - Detected tablet split for table ks.test, increasing from 1 to 2 tablets
INFO  2024-10-22 20:59:26,470 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1155828
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Emitting resize decision of type split for table 2984a460-908a-11ef-93f4-09ea64b31b3e due to avg tablet size of 1155828
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:59:26,471 [shard 0: gms] raft_topology - updating topology state: Starting tablet migration
INFO  2024-10-22 20:59:26,478 [shard 0: gms] raft_topology - entered `tablet migration` transition state
INFO  2024-10-22 20:59:26,478 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,478 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 20:59:26,479 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,479 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1155828
INFO  2024-10-22 20:59:26,479 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 20:59:26,479 [shard 0: gms] raft_topology - updating topology state: Finished tablet migration
INFO  2024-10-22 20:59:26,479 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_3nlnk28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 20:59:26,479 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_2unzk28zy0sfxww666-big-Data.db` : 12882
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1155828
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
WARN  2024-10-22 20:59:26,559 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_3bt9s28zy0sfxww666-big-Data.db` : 12882
INFO  2024-10-22 20:59:26,677 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_2unzk28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_3bt9s28zy0sfxww666-big-Data.db:level=0]. 1MB to 1MB (~100% of original) in 194ms = 5MB/s. ~24960 total partitions merged to 24937.
INFO  2024-10-22 20:59:26,678 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 20:59:26,678 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_41bhc28zy0sfxww666-big-Data.db` : 12948
WARN  2024-10-22 20:59:26,780 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db` : 12948
INFO  2024-10-22 20:59:26,876 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_41bhc28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db:level=0]. 1MB to 1MB (~100% of original) in 195ms = 5MB/s. ~25088 total partitions merged to 25062.
INFO  2024-10-22 20:59:26,877 [shard 0:strm] table - Setting split ready sequence number to 3 for table ks.test
INFO  2024-10-22 21:00:26,438 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1161101
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 3
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:00:26,439 [shard 0: gms] raft_topology - updating topology state: Started tablet split finalization
INFO  2024-10-22 21:00:26,455 [shard 0: gms] raft_topology - entered `tablet split finalization` transition state
INFO  2024-10-22 21:00:26,456 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
INFO  2024-10-22 21:00:26,456 [shard 0: gms] raft_topology - updating topology state: advance fence version to 10
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1161101
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 3
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Split tablets for table 2984a460-908a-11ef-93f4-09ea64b31b3e, increasing tablet count from 2 to 4
INFO  2024-10-22 21:00:26,464 [shard 0: gms] raft_topology - updating topology state: Finished tablet split finalization
INFO  2024-10-22 21:00:26,470 [shard 0: gms] table - Detected tablet split for table ks.test, increasing from 2 to 4 tablets
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 580550
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Emitting resize decision of type split for table 2984a460-908a-11ef-93f4-09ea64b31b3e due to avg tablet size of 580550
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:00:26,472 [shard 0: gms] raft_topology - updating topology state: Starting tablet migration
INFO  2024-10-22 21:00:26,479 [shard 0: gms] raft_topology - entered `tablet migration` transition state
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 580550
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 21:00:26,479 [shard 0: gms] raft_topology - updating topology state: Finished tablet migration
INFO  2024-10-22 21:00:26,480 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_41bhc28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,480 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_2uvpc28zy0sfxww666-big-Data.db` : 6208
INFO  2024-10-22 21:00:26,488 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,488 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:00:26,489 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,489 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 580550
INFO  2024-10-22 21:00:26,489 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 21:00:26,526 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_2uvpc28zy0sfxww666-big-Filter.db: resizing bitset from 7768 bytes to 7856 bytes. sstable origin: split
WARN  2024-10-22 21:00:26,536 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_36vsw28zy0sfxww666-big-Data.db` : 6984
INFO  2024-10-22 21:00:26,574 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_36vsw28zy0sfxww666-big-Filter.db: resizing bitset from 8736 bytes to 7600 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,587 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_2uvpc28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_36vsw28zy0sfxww666-big-Data.db:level=0]. 573kB to 577kB (~100% of original) in 104ms = 5MB/s. ~12416 total partitions merged to 12350.
INFO  2024-10-22 21:00:26,587 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_3bt9s28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,588 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3i11c28zy0sfxww666-big-Data.db` : 6208
WARN  2024-10-22 21:00:26,631 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3r8ts28zy0sfxww666-big-Data.db` : 6984
INFO  2024-10-22 21:00:26,671 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3r8ts28zy0sfxww666-big-Filter.db: resizing bitset from 8736 bytes to 7848 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,686 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3i11c28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3r8ts28zy0sfxww666-big-Data.db:level=0]. 576kB to 580kB (~100% of original) in 95ms = 6MB/s. ~12416 total partitions merged to 12402.
INFO  2024-10-22 21:00:26,687 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_2unzk28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,687 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_438xe28zy0sfxww666-big-Data.db` : 6272
WARN  2024-10-22 21:00:26,734 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4dbkw28zy0sfxww666-big-Data.db` : 7056
INFO  2024-10-22 21:00:26,771 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4dbkw28zy0sfxww666-big-Filter.db: resizing bitset from 8824 bytes to 7960 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,787 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_438xe28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4dbkw28zy0sfxww666-big-Data.db:level=0]. 582kB to 586kB (~100% of original) in 97ms = 6MB/s. ~12544 total partitions merged to 12535.
INFO  2024-10-22 21:00:26,787 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,788 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4ow8w28zy0sfxww666-big-Data.db` : 6400
INFO  2024-10-22 21:00:26,826 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4ow8w28zy0sfxww666-big-Filter.db: resizing bitset from 8008 bytes to 8024 bytes. sstable origin: split
WARN  2024-10-22 21:00:26,837 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4zec028zy0sfxww666-big-Data.db` : 7200
INFO  2024-10-22 21:00:26,877 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4zec028zy0sfxww666-big-Filter.db: resizing bitset from 9008 bytes to 7880 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,890 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4ow8w28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4zec028zy0sfxww666-big-Data.db:level=0]. 590kB to 594kB (~100% of original) in 99ms = 5MB/s. ~12800 total partitions merged to 12712.
INFO  2024-10-22 21:00:26,891 [shard 0:strm] table - Setting split ready sequence number to 5 for table ks.test
INFO  2024-10-22 21:01:26,438 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 585003
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 5
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:01:26,439 [shard 0: gms] raft_topology - updating topology state: Started tablet split finalization
INFO  2024-10-22 21:01:26,456 [shard 0: gms] raft_topology - entered `tablet split finalization` transition state
INFO  2024-10-22 21:01:26,456 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
INFO  2024-10-22 21:01:26,456 [shard 0: gms] raft_topology - updating topology state: advance fence version to 14
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 585003
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 5
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Split tablets for table 2984a460-908a-11ef-93f4-09ea64b31b3e, increasing tablet count from 4 to 8
INFO  2024-10-22 21:01:26,465 [shard 0: gms] raft_topology - updating topology state: Finished tablet split finalization
INFO  2024-10-22 21:01:26,471 [shard 0: gms] table - Detected tablet split for table ks.test, increasing from 4 to 8 tablets
INFO  2024-10-22 21:01:26,472 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:01:26,472 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=8, tablets=8, shards=1, state=normal
INFO  2024-10-22 21:01:26,472 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
```

Reply:
The patch didn't help if the partitions were too large. I tried with few large partitions, each having rows in the range of [10, 100000] and with the patch, after a split most of the rebuilds were still happening.

Reply:
> - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db` : 12948

I can think of index sampling level (summary) having an impact, but with large partitions, I'd expect each large partition to have its own summary entry to due to data:summary byte ratio of 2000:1.

Reply:
One possible reason: 

The logic that decides if a filter has to be rebuilt, skips rebuilding if the false positive rate is within +/-25% of the configured fp rate. For small Bloom filters, even a minor variation in the estimate can easily push the false positive rate beyond the allowed limit, triggering rebuilds. 

I tested with 100 large partitions, and for 100 keys, the bloom filter size is already small : 128 bytes. When split, the bloom filter sizes keep decreasing.  Maybe that is why I'm seeing many rebuilds.